Token automatic refresh

patpicos opened the pull request:
Leaving the token refresh up to the user is likely to lead to a token expiring and calls to TV DB failing. I added a channel with a timer sending events every 6 hours to update `c.token`

This has a breaking change where I hid RefreshToken() because it now iterates over the channel. so the method should not be called externally going forward.

I am not sure how to add a test to evaluate this. I would have liked the channel to be part of the struct, but that would require a change to the tvdb.Client() to have the developer create the dependency and inject it (this would be better because it would open up the ability to mock a shorter refresh timer and compare c.token (pre refresh) and c.token (post refresh).

This PR is to open the discussion on the changes. Can modify once we come to an agreement :)

patpicos committed:
Merge pull request #1 from pioz/master

sync

patpicos committed:
added token auto-refresh  every 6 hours

pioz commented on client.go:
```diff
@@ -54,22 +57,27 @@ func (c *Client) Login() error {
-	if err != nil {
-		return err
+// refreshToken is used to refresh the current token.
+func (c *Client) refreshToken() error {
```
I would keep this method public so a user can manage the token refresh also manually. This also does not introduce any breaking changes.

pioz commented on client.go:
```diff
@@ -54,22 +57,27 @@ func (c *Client) Login() error {
 	}
 	c.token = data.Token
+
+	go c.refreshToken()
```
Change with something like this.
```
go func() {
  for range tickCh {
    c.RefreshToken()
  }
}()

pioz commented on client.go:
```diff
@@ -54,22 +57,27 @@ func (c *Client) Login() error {
+		if err != nil {
+			return err
+		}
+		c.token = data.Token
```
May be we need to add a lock to prevent race conditions because `RefreshToken` is public?

pioz commented on client_test.go:
```diff
@@ -22,13 +22,13 @@ func TestClientLoginFail(t *testing.T) {
-		t.Fatal(err)
-	}
-}
+// func TestClientRefreshToken(t *testing.T) {
```
Uncomment this.

pioz commented on client.go:
```diff
@@ -40,6 +41,8 @@ type Client struct {
 // BaseURL where the TVDB api is accessible.
 const BaseURL string = "https://api.thetvdb.com"
 
+var tickCh = time.Tick(6 * time.Hour)
```
What about adding a public method to change the duration? Something like:
```
func SetAutoRefreshTokenEvery(d time.Duration)
```

pioz reviewed (changes requested):
I would keep the `RefreshToken` method public so a user can manage the token refresh also manually as he want. This also does not introduce any breaking changes.

I like the feature to autorefresh the token!

What do you think about my review?

patpicos commented:
> I would keep the `RefreshToken` method public so a user can manage the token refresh also manually as he want. This also does not introduce any breaking changes.
> 
> I like the feature to autorefresh the token!
> 
> What do you think about my review?

Makes sense. I plan on making the auto-refresh per client rather than package level.

patpicos committed:
Updated auto-refresh

patpicos commented:
Hey @pioz please have a look.

I am having issues with validating the test `TestSetAutoRefreshTokenEvery()`. I force the refresh to be every 3 second and want to compare the initial token with the "new" one and validate they are different. However, the `c.token` doesnt seem to stick. See the logging i added.

Example: (shortened the outputs since they are sensitive)
```
patrick@DESKTOP-FIGGPSQ:~/go/src/github.com/pioz/tvdb$ go test -timeout 90m github.com/pioz/tvdb -run ^TestSetAutoRefreshTokenEvery$ -v -count 1
=== RUN   TestSetAutoRefreshTokenEvery
    TestSetAutoRefreshTokenEvery: client_test.go:37: Initial token  EBqwAUw
2020/08/09 09:51:06 Token before:  EBqwAUw
2020/08/09 09:51:06 Token after:  SAsAKSg //Shows that the struct is updating
2020/08/09 09:51:09 Token before:  SAsAKSg
2020/08/09 09:51:09 Token after:  g3-z9oXQ //Shows that the struct is updating
2020/08/09 09:51:12 Token before:  g3-z9oXQ
2020/08/09 09:51:12 Token after:  VA5vhUcJIIw //Shows that the struct is updating
    TestSetAutoRefreshTokenEvery: client_test.go:43: Updated token  EBqwAUw  //Same as the initial. Expecting to get `VA5vhUcJIIw `
--- PASS: TestSetAutoRefreshTokenEvery (10.44s)
PASS
ok      github.com/pioz/tvdb    10.452s

```

At least it shows the ticker is running on an updated schedule.

patpicos commented:
did more investigation and it feels like the goroutine are working on a different object (logs do show an updating version, but not seen externally).

I went back to my original implementation...and the same occurs. I am at a loss at how to make this work (this is my first attempt at concurrent golang!)
